ui/app/src/Views/AlertList/Updates.elm

```diff
@@ -1,33 +1,76 @@
 module Views.AlertList.Updates exposing (..)
 
 import Alerts.Api as Api
-import Views.AlertList.Types exposing (AlertListMsg(..), Model)
+import Views.AlertList.Types exposing (AlertListMsg(..), Model, Tab(FilterTab, GroupTab))
 import Views.FilterBar.Updates as FilterBar
-import Navigation
 import Utils.Filter exposing (Filter, parseFilter)
-import Utils.Types exposing (ApiData, ApiResponse(Loading))
+import Utils.Types exposing (ApiData, ApiResponse(Initial, Loading, Success, Failure))
 import Types exposing (Msg(MsgForAlertList, Noop))
-import Dom
-import Task
+import Set
+import Navigation
+import Utils.Filter exposing (generateQueryString)
+import Views.GroupBar.Updates as GroupBar
 
 
 update : AlertListMsg -> Model -> Filter -> ( Model, Cmd Types.Msg )
-update msg model filter =
+update msg ({ groupBar, filterBar } as model) filter =
     case msg of
         AlertsFetched listOfAlerts ->
-            ( { model | alerts = listOfAlerts }, Cmd.none )
-
-        FetchAlerts ->
             ( { model
-                | filterBar = FilterBar.setMatchers filter model.filterBar
-                , alerts = Loading
+                | alerts = listOfAlerts
+                , groupBar =
+                    case listOfAlerts of
+                        Success alerts ->
+                            { groupBar
+                                | list =
+                                    List.concatMap .labels alerts
+                                        |> List.map Tuple.first
+                                        |> Set.fromList
+                            }
+
+                        _ ->
+                            groupBar
               }
-            , Api.fetchAlerts filter |> Cmd.map (AlertsFetched >> MsgForAlertList)
+            , Cmd.none
             )
 
+        FetchAlerts ->
+            let
+                newGroupBar =
+                    GroupBar.setFields filter groupBar
+
+                newFilterBar =
+                    FilterBar.setMatchers filter filterBar
+
+                newModel =
+                    { model | filterBar = newFilterBar, groupBar = newGroupBar }
+            in
+                -- only fetch when filter changed
+                if newFilterBar /= model.filterBar || model.alerts == Initial then
```

Comment:
This appears to be preventing my UI from re-querying for alerts when I remove a label matcher.

Trying to be efficient is cool, but I really don't think it's needed here. The actual load on any one alertmanager should be pretty small, and users should be round-robin'ing through their HA alertmanager instances. It seems that this new `Initial` type exists for the same reason? I would vote for removing this as I don't think the small increase in performance is worth adding the extra type and then having to worry about the sequence of events leading to the `FetchAlerts` message, as it's already caused a bug.